tmva/tmva/inc/TMVA/DNN/SGD.h

```diff
@@ -0,0 +1,171 @@
+
+   /*! Update the biases, given the current bias gradients. */
+   void UpdateBiases(size_t layerIndex, std::vector<Matrix_t> &biases, const std::vector<Matrix_t> &biasGradients);
+
+public:
+   /*! Constructor. */
+   TSGD(Scalar_t learningRate, DeepNet_t &deepNet, Scalar_t momentum);
+
+   /*! Destructor. */
+   ~TSGD();
+
+   /*! Getters */
+   inline Scalar_t GetMomentum() const { return fMomentum; }
+
+   inline std::vector<std::vector<Matrix_t>> GetPastWeightGradients() { return fPastWeightGradients; }
+   inline std::vector<Matrix_t> GetPastWeightGradientsAt(size_t i) { return fPastWeightGradients[i]; }
+
+   inline std::vector<std::vector<Matrix_t>> GetPastBiasGradients() { return fPastBiasGradients; }
+   inline std::vector<Matrix_t> GetPastBiasGradientsAt(size_t i) { return fPastBiasGradients[i]; }
+};
+
+//
+//
+//  The Stochastic Gradient Descent Optimizer Class - Implementation
+//_________________________________________________________________________________________________
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+TSGD<Architecture_t, Layer_t, DeepNet_t>::TSGD(Scalar_t learningRate, DeepNet_t &deepNet, Scalar_t momentum)
+   : VOptimizer<Architecture_t, Layer_t, DeepNet_t>(learningRate, deepNet), fMomentum(momentum)
+{
+   int layersNSlices = deepNet.GetLayers().size();
+   fPastWeightGradients.resize(layersNSlices);
+   fPastBiasGradients.resize(layersNSlices);
+
+   for (int i = 0; i < layersNSlices; i++) {
+      int weightsNSlices = (deepNet.GetLayerAt(i)->GetWeights()).size();
+
+      for (int j = 0; j < weightsNSlices; j++) {
+         int weightsNRows = (deepNet.GetLayerAt(i)->GetWeightsAt(j)).GetNrows();
+         int weightsNCols = (deepNet.GetLayerAt(i)->GetWeightsAt(j)).GetNcols();
+
+         fPastWeightGradients[i].emplace_back(weightsNRows, weightsNCols);
+         initialize<Architecture_t>(fPastWeightGradients[i][j], EInitialization::kZero);
+      }
+
+      int biasesNSlices = (deepNet.GetLayerAt(i)->GetBiases()).size();
+
+      for (int j = 0; j < biasesNSlices; j++) {
+         int biasesNRows = (deepNet.GetLayerAt(i)->GetBiasesAt(j)).GetNrows();
+         int biasesNCols = (deepNet.GetLayerAt(i)->GetBiasesAt(j)).GetNcols();
+
+         fPastBiasGradients[i].emplace_back(biasesNRows, biasesNCols);
+         initialize<Architecture_t>(fPastBiasGradients[i][j], EInitialization::kZero);
+      }
+   }
+}
+
+//_________________________________________________________________________________________________
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+TSGD<Architecture_t, Layer_t, DeepNet_t>::~TSGD()
+{
+}
+
+//_________________________________________________________________________________________________
+template <typename Architecture_t, typename Layer_t, typename DeepNet_t>
+auto TSGD<Architecture_t, Layer_t, DeepNet_t>::UpdateWeights(size_t layerIndex, std::vector<Matrix_t> &weights,
```

Comment:
Since all the methods in other classes ( like DenseLayer, GeneralLayer classes )  were using a trailing return type ( even for void return type ). So I used them to ensure consistency. I can change that if needed?

Reply:
OK fine. Consistency wins. It's possibly bad consistency though - but not for you to fix!